### PR TITLE
Description version

### PR DESCRIPTION
**Pull request by sahithyaravi**
### Adding version for descriptions:

**Approach:**
Created a new table which stores descriptions. The primary key of the description table = did and version(description version).

**APIs modified as a result:**
- _data_upload_ - Store description in a new table with default version number 1
- _data_edit_ - adds a new version to description table for that did
- _data_ (get data) - gets latest version for that did from description table

### CODE REVIEW

File: downloads/openml_expdb.sql [hunk 1]
```diff
@@ -148,6 +147,20 @@ CREATE TABLE `dataset_tag` (
 -- --------------------------------------------------------
 
 
+--
+-- Table structure for table `dataset_description`
+--
+
+CREATE TABLE `dataset_description` (
+  `did` int(10) UNSIGNED NOT NULL,
+  `version` varchar(64) COLLATE utf8_unicode_ci NOT NULL,
```
Comment (janvanrijn) [inline downloads/openml_expdb.sql]: probably better to make this a date time, instead of a integer?

File: openml_OS/models/api/v1/Api_data.php [hunk 2]
```diff
@@ -589,18 +590,34 @@ private function data_edit() {
...
-      $this->returnError( 1067, $this->version );
-      return;
+    // Add description in the description table as a new version
+    if (isset($update_fields['description']))
+    {
+      $description_record = $this->Dataset_description->getWhereSingle('did =' . $data_id, 'version DESC');
+      $version_new = $description_record->version + 1;
+
+      $desc = array(
+          'did'=>$data_id,
```
Comment (janvanrijn) [inline openml_OS/models/api/v1/Api_data.php]: please consistently use 2 space indention

File: openml_OS/models/api/v1/Api_data.php [hunk 3]
```diff
@@ -589,18 +590,34 @@ private function data_edit() {
...
     }
 
-    $update_result = $this->Dataset->update($data_id, $update_fields);
-    // If result returns error    
-    if( $update_result == false ) {
-      $this->returnError( 1067, $this->version );
-      return;
+    // Add description in the description table as a new version
+    if (isset($update_fields['description']))
+    {
```
Comment (janvanrijn) [inline openml_OS/models/api/v1/Api_data.php]: please do accolades on previous line

File: openml_OS/models/api/v1/Api_data.php [hunk 4]
```diff
@@ -589,18 +590,34 @@ private function data_edit() {
...
-    try {
-      $this->elasticsearch->index('data', $data_id);
-    } catch (Exception $e) {
-      $this->returnError(105, $this->version, $this->openmlGeneralErrorCode, $e->getMessage());
+
+    if ($update_fields)
+    {
+      $update_result = $this->Dataset->update($data_id, $update_fields);
+      // If result returns error
+      if( $update_result == false ) {
```
Comment (janvanrijn) [inline openml_OS/models/api/v1/Api_data.php]: Try to keep the convention:
``if ($update_result == false) {``

File: openml_OS/models/api/v1/Api_data.php [hunk 5]
```diff
@@ -589,18 +590,34 @@ private function data_edit() {
...
     }
-    
-    // update elastic search index.  
-    try {
-      $this->elasticsearch->index('data', $data_id);
-    } catch (Exception $e) {
-      $this->returnError(105, $this->version, $this->openmlGeneralErrorCode, $e->getMessage());
+
+    if ($update_fields)
+    {
```
Comment (janvanrijn) [inline openml_OS/models/api/v1/Api_data.php]: prev line

File: openml_OS/models/api/v1/Api_data.php [hunk 6]
```diff
@@ -589,18 +590,34 @@ private function data_edit() {
...
+      unset($update_fields['description']);
+      $desc_id = $this->Dataset_description->insert($desc);
+      if (!$desc_id) {
+        $this->returnError(1067, $this->version);
+        return;
+      }
     }
-    
-    // update elastic search index.  
-    try {
```
Comment (janvanrijn) [inline openml_OS/models/api/v1/Api_data.php:600]: i like that this is removed!! but where does it go?

File: openml_OS/models/api/v1/Api_data.php [hunk 7]
```diff
@@ -1118,6 +1138,15 @@ private function data_upload() {
       unset($dataset['tag']);
     }
 
+
+    $desc = array(
+        'version' => 1,
```
Comment (janvanrijn) [inline openml_OS/models/api/v1/Api_data.php]: 2 spaces

File: openml_OS/models/api/v1/Api_data.php (hunk 6 above)
Comment (sahithyaravi) [reply openml_OS/models/api/v1/Api_data.php:600]: Sorry this was a merge mistake. I should not be removing this. (Will fix it with other changes)

File: downloads/openml_expdb.sql (hunk 1 above)
Comment (sahithyaravi) [reply downloads/openml_expdb.sql]: I used version numbers similar to the way it is done for dataset versioning. Do you think using an integer may cause problems if there are concurrent "Edits"?

Comment (janvanrijn) [reply downloads/openml_expdb.sql]: I think that `did` is the primary key of dataset, and `name`, `version` is there an alternative key, that gives users the option to specify dataset versions in a non-numeric way. 

I do not think we need that for the description version. `did` maps as foreign key to the dataset table, and version can just increment numerically (or according to some other logical scheme, such as a date).

Comment (janvanrijn) [reply downloads/openml_expdb.sql]: Concurrent edits are indeed a problem, with e.g., knowing which edit was the most recent and which you want to present to the user. How do you want to solve that?

(Separate question. I don't think that having a varchar data field solves the issue. )

Comment (sahithyaravi) [reply downloads/openml_expdb.sql]: Could we just use datetime with a resolution of milli or microseconds.
I think concurrent updates would still differ in milli seconds if not seconds? Something like:
"2012-07-08 11:14:15.889342"

Comment (janvanrijn) [reply downloads/openml_expdb.sql]: I mean the problem is when two persons make an adjustment based on text A. Person 1 will maybe make a big update, whereas person 2 will improve some typos. If both are based upon version A, then the edits of person 1 will be stowed away in the database. But maybe that's something we have to accept.

I wouldn't be to worried about the database overwriting a certain entry, that could be easily avoided by relying on mysqls internal procedures.

Comment (sahithyaravi) [reply downloads/openml_expdb.sql]: Yes, may be we could accept this behavior as it may be rare as well.

Comment (sahithyaravi) [reply downloads/openml_expdb.sql]: "version can just increment numerically (or according to some other logical scheme, such as a date"

So we are incrementing the version numerically currently. I think we could go ahead with this, unless there is another reason to switch to date. We could accept that in concurrent edits there is a possibility that the database may reject a change and pick another(as simultaneous edits may be a rare occurrence)

**Comment by joaquinvanschoren** [comment]
Could we do a simple check: in case there is a version conflict, return a
special error message saying “the description on the server has changed,
please make changes to the new version”. Git does something similar. The
clients could still do an automated merge if possible, but likely the user
has to check it anyway.

File: downloads/openml_expdb.sql (hunk 1 above)
Comment (sahithyaravi) [reply downloads/openml_expdb.sql]: @janvanrijn I will create an API list_description_version(did) which can list all  versions of a description. This way it is possible to get the history and restore to an earlier version if something goes wrong.

**Review by janvanrijn — APPROVED** [review]

**Pull request merged**